sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc

```diff
@@ -0,0 +1,200 @@
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2024-2025 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+Nvidia's compute capability (cc) 9.0 devices introduced a new level in the thread hierarchy, called as thread block clusters, in CUDA terminology. A thread block cluster, is a collection of thread blocks (a workgroup in SYCL terminology), which are guaranteed to be scheduled to be scheduled on a separate compute unit, and are guaranteed to run in parallel, and the thread blocks in the cluster can access each other's shared memory (local memory SYCL terminology). This has various applications, convolutions, GEMMs and FFTs to name a few.
+
+This proposal introduces a SYCL equivalent of the same, to be known as `cluster_group`, and adds the methods to launch a kernel with the cluster size mentioned, accessing the id's of the workgroup in the cluster, and to access another workgroup's local memory within the `cluster_group`.
+
+
+== Specification
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_CLUSTER_GROUP` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Launching a kernel with work group clusters
+
+To launch a kernel with a cluster dimension, because of the special scheduling guarantees, the driver requires which kernel will be using the said feature. This introduces an additional parallel_for overload and parallel_for_work_group overloads, accepting cluster dimensions as a `sycl::range`, the dimension of which will be the same as that of the global and the local range.
```

Comment:
I think it would also make sense to have an aspect for querying whether a device supports the cluster-group launch, though since this is an Nvidia-specific extension, this may not be strictly necessary atm. @gmlueck what are you thoughts on this?

Reply:
> Using enqueue functions and properties now,
Though if @AerialMantis or maybe @gmlueck , if you could please once verify if simply defining a subscript operator would be okay to return the property values / or it's usage in general, it would be really helpful.

I think having a subscript operator on the property would be useful, and I think there is precident for this in other extensions, see the `work_group_size` property in [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc).

> I was planning on using sycl_ext_oneapi_device_architecture to check for CC 9.0, but just wanted to ask for other's opinion before committing to it. For now, I have not mentioned the behaviour of the backend in case this is used on unsupported devices, and can be implementation specific

I think it makes sense to check the compute capability as a way for the backend to verify that it can support a cluster group launch, though despite this being a CUDA-specific extension, I would reflect this as an aspect, and leave the architecture check as an implementation detail.

> On a similar topic, should we also extend get_info on what are the valid ranges for the cluster dimension ?

I think this is a good idea, it would be useful for users to be able to query what the maximum cluster group size can be, something like `info::device::max_cluster_group_size`. I wonder if it's possible for the maximum cluster group size to change based on the compilation of the kernel, in which case we may want to have a compiled kernel query as well, something like `info::kernel_device_specific::max_cluster_group_size`. In CUDA `cudaOccupancyMaxPotentialClusterSize` takes the kernel function as a parameter so that would suggest that this may be the case.

Reply:
Added the device query `ext_cuda_experimental_max_cluster_group_size`
following the naming convention for additional device queries.